php/class-customize-snapshot-manager-back-compat.php

```diff
@@ -0,0 +1,1009 @@
+				}
+			}
+		}
+
+		/*
+		 * Disable routine which fails because \WP_Customize_Manager::setup_theme() is
+		 * never called in a frontend preview context, whereby the original_stylesheet
+		 * is never set and so \WP_Customize_Manager::is_theme_active() will thus
+		 * always return true because get_stylesheet() !== null.
+		 *
+		 * The action being removed is responsible for adding an option_sidebar_widgets
+		 * filter \WP_Customize_Widgets::filter_option_sidebars_widgets_for_theme_switch()
+		 * which causes the sidebars_widgets to be overridden with a global variable.
+		 */
+		if ( ! is_admin() ) {
+			remove_action( 'wp_loaded', array( $this->customize_manager->widgets, 'override_sidebars_widgets_for_theme_switch' ) );
+		}
+	}
+
+	/**
+	 * Add filters for previewing nav menus on the frontend.
+	 */
+	public function add_nav_menu_setting_preview_filters() {
+		if ( isset( $this->customize_manager->nav_menus ) && ! current_user_can( 'edit_theme_options' ) ) {
+			$hooks = array(
+				'customize_register' => array(
+					'callback' => array( $this->customize_manager->nav_menus, 'customize_register' ),
+					'priority' => 11,
+				),
+				'customize_dynamic_setting_args' => array(
+					'callback' => array( $this->customize_manager->nav_menus, 'filter_dynamic_setting_args' ),
+					'priority' => 10,
+				),
+				'customize_dynamic_setting_class' => array(
+					'callback' => array( $this->customize_manager->nav_menus, 'filter_dynamic_setting_class' ),
+					'priority' => 10,
+				),
+				'wp_nav_menu_args' => array(
+					'callback' => array( $this->customize_manager->nav_menus, 'filter_wp_nav_menu_args' ),
+					'priority' => 1000,
+				),
+				'wp_nav_menu' => array(
+					'callback' => array( $this->customize_manager->nav_menus, 'filter_wp_nav_menu' ),
+					'priority' => 10,
+				),
+			);
+			foreach ( $hooks as $hook_name => $hook_args ) {
+				// Note that add_action()/has_action() are just aliases for add_filter()/has_filter().
+				if ( ! has_filter( $hook_name, $hook_args['callback'] ) ) {
+					add_filter( $hook_name, $hook_args['callback'], $hook_args['priority'], PHP_INT_MAX );
+				}
+			}
+		}
+
+		if ( isset( $this->customize_manager->nav_menus ) ) {
+			add_action( 'customize_register', array( $this, 'preview_early_nav_menus_in_customizer' ), 9 );
+		}
+	}
+
+	/**
+	 * Preview nav menu settings early so that the sections and controls for snapshot values will be added properly.
+	 *
+	 * This must happen at `customize_register` priority prior to 11 which is when `WP_Customize_Nav_Menus::customize_register()` runs.
+	 * This is only relevant when accessing the Customizer app (customize.php), as this is where sections/controls matter.
+	 *
+	 * @see \WP_Customize_Nav_Menus::customize_register()
+	 */
+	public function preview_early_nav_menus_in_customizer() {
+		if ( ! is_admin() ) {
+			return;
+		}
+		$this->customize_manager->add_dynamic_settings( array_keys( $this->snapshot()->data() ) );
```

Comment:
This `Customize_Snapshot_Manager_Back_Compat` should probably override the `snapshot` method to just indicate that it returns a `Customize_Snapshot_Back_Compat`, since the `data` method is only available in that subclass. Or implement the `data` method on the base class.